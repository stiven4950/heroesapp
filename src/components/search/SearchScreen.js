import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';

import { useForm } from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

const SearchScreen = ({history}) => {

    // Para recibir de forma más sencilla los parámetros de los query string
    // se instala un paquete que hace eso por nosotros.
    // npm install query-string

    const { search } = useLocation();
    const { q = '' } = queryString.parse(search);

    
    const [{ description }, handleInputChange] = useForm({
        description: q,
    });
    
    // Se está disparando en cada instante se resuelve con useMemo
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();

        history.push(`?q=${description}`);
    }

    return (
        <>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            name="description"
                            value={description}
                            className="form-control"
                            placeholder="Find your hero"
                            onChange={handleInputChange}
                            autoComplete="off"
                            required
                        />
                        <div className="d-grid gap-2">
                            <button
                                type="submit"
                                className="btn m-1 btn-outline-primary btn-block"
                            >
                                Search
                            </button>
                        </div>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '')
                        &&
                        <div className="alert alert-info">
                            Search a hero
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltered.length === 0)
                        &&
                        <div className="alert alert-danger">
                            There is no a hero with <b>{q}</b>
                        </div>
                    }
                    {heroesFiltered.map(heroe =>
                        <HeroCard
                            key={heroe.id}
                            {...heroe}
                        />
                    )}
                </div>
            </div>
        </>
    );
}

export default SearchScreen;