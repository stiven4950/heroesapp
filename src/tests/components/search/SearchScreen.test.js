import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import SearchScreen from '../../../components/search/SearchScreen';

describe('Pruebas en el componente <SearchScreen />', () => {
    test('Debe de mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    test('Debe de mostrar a batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar un error si no se muestra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman1235']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no a hero with batman1235');
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar el push del history', () => {
        const history = {
            push: jest.fn(),
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman1235']}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'description',
                value: 'batman',
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault() { },
        });

        expect(history.push).toHaveBeenCalledWith('?q=batman');
    });


});