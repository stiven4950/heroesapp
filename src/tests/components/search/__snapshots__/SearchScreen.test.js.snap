// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrar a batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={[Function]}
          placeholder="Find your hero"
          required={true}
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn m-1 btn-outline-primary btn-block"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 340,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <div
                className="card-title"
              >
                Batman
              </div>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrar un error si no se muestra el hero 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={[Function]}
          placeholder="Find your hero"
          required={true}
          type="text"
          value="batman1235"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn m-1 btn-outline-primary btn-block"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with 
        <b>
          batman1235
        </b>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={[Function]}
          placeholder="Find your hero"
          required={true}
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn m-1 btn-outline-primary btn-block"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>,
]
`;
